cpp-client/deephaven/client/src/server/server.cc

```diff
@@ -75,26 +75,68 @@ const char *timeoutKey = "http.session.durationMs";
 const size_t handshakeResendIntervalMillis = 5 * 1000;
 }  // namespace
 
-std::shared_ptr<Server> Server::createFromTarget(const std::string &target, const std::string &authorizationValue) {
-  auto channel = grpc::CreateChannel(target, grpc::InsecureChannelCredentials());
+namespace {
+std::shared_ptr<grpc::ChannelCredentials> get_credentials(const bool useTls, const std::string &pem) {
+  if (!useTls) {
+    return grpc::InsecureChannelCredentials();
+  }
+  grpc::SslCredentialsOptions options;
+  if (pem != "") {
+    options.pem_root_certs = pem;
+  }
+  return grpc::SslCredentials(options);
+}
+}  // namespace
+
+std::shared_ptr<Server> Server::createFromTarget(
+      const std::string &target,
+      const std::string &authorizationValue) {
+  return createFromTarget(target, authorizationValue, ClientOptions::defaults());
+}
+
+std::shared_ptr<Server> Server::createFromTarget(
+      const std::string &target,
+      const std::string &authorizationValue,
+      const ClientOptions &copts) {
+  if (!copts.useTls() && copts.pem() != "") {
+    throw std::runtime_error(
+        "Server::createFromTarget: ClientOptions: useTls is false but pem provided");
+  }
+
+  grpc::ChannelArguments channel_args;
+  for (const std::pair<std::string, int>& opt : copts.intOptions()) {
+    channel_args.SetInt(opt.first, opt.second);
+  }
+  for (const std::pair<std::string, std::string>& opt : copts.stringOptions()) {
+    channel_args.SetString(opt.first, opt.second);
+  }
+
+  auto channel = grpc::CreateCustomChannel(
+      target, get_credentials(copts.useTls(), copts.pem()), channel_args);
   auto as = ApplicationService::NewStub(channel);
   auto cs = ConsoleService::NewStub(channel);
   auto ss = SessionService::NewStub(channel);
   auto ts = TableService::NewStub(channel);
   auto cfs = ConfigService::NewStub(channel);
 
   // TODO(kosak): Warn about this string conversion or do something more general.
-  auto flightTarget = "grpc://" + target;
+  auto flightTarget = ((copts.useTls()) ? "grpc+tls://" : "grpc://") + target;
   arrow::flight::Location location;
 
   auto rc1 = arrow::flight::Location::Parse(flightTarget, &location);
   if (!rc1.ok()) {
-    auto message = stringf("Location::Parse(%o) failed, error = %o", flightTarget, rc1.ToString());
+    auto message = stringf("Location::Parse(%o) failed, error = %o",
+                           flightTarget, rc1.ToString());
     throw std::runtime_error(message);
```

Comment:
```
throw std::runtime_error(DEEPHAVEN_DEBUG_MSG(message));
```

Reply:
We don't consistently use DEEPHAVEN_DEBUG_MSG everywhere else in this file (ie, we were throwing in other cases without it), I looked up the definition and tried to deduce a rationale for when to use it, I guess I deduced wrong?

Reply:
Yeah. I would recommend using it on every single throw. It's definitely not done in a lot of places. But, anywhere it isn't being done is an oversight. :-)

Reply:
Understood.